Adapt GOOPy to support Jacobian simulations

hannahnesser opened the pull request:
There are a series of changes in this pull request, though most of them center around improved efficiency for the use case of applying GOOPy to Jacobian simulations (for application in the IMI). In these simulations, GEOS-Chem is run 100+ times with slightly different fluxes and the atmosphere is sampled in the same way in each simulation. They also often use multiple tracers (e.g., CH4_0001, CH4_0002) within one simulation for computational improvement.  This pull request includes supporting changes, mostly in two categories: 

1. We now allow some of the key outputs of the operator (the interpolation matrices and space-time mapping indices) to be pre-generated. We add a config option SAVE_INTERPOLATION accordingly. If this is true, the operator will save these files out in the observation directory (in a sub directory called "operator_components"). These files can be large. If this is false, the operator will not save these files out. In both cases, the operator will attempt to load these files. If it finds the files, it will use them. If not, it will re-calculate it. In practice, this means that for standard (non-Jacobian) applications, SAVE_INTERPOLATION should be False. For Jacobian applications, SAVE_INTERPOLATION will be True for the prior simulation and False for subsequent simulations.

2. We modify the assumption that the GEOS-Chem files only have one species of interest. This requires changing the parsers for GEOS-Chem and the assumed dimensions of the interpolation operators, etc.

Commit:
Adding placeholders to the config file that are compatible with an IMI implementation of GOOPy; also adding the option to SAVE_INTERPOLATION for Jacobian runs, which allows interpolation operators to be re-used for different GEOS-Chem simulations using the same observations

Commit:
A variety of changes
- Added number of threads arguments to improve performance (needs to be checked that this actually helps)
- Added save_interpolation option to support application of GOOPy to Jacobian simulations (also in the config file)
- Added save_dir argument for GOOPy outputs (also in the config file)
- Modified the dimension checks to account for Jacobian simulations (since we are often processing GEOS-Chem files that have multiple species)
- Allow for loading OR computing the interpolation map, depending on whether the interpolation files exist in the OBS_DIR/operator_files directory

Commit:
Removing timing statements

Commit:
- Added num_threads arguments to try to improve performance (this is relatively untested)
- Changed .dims to .sizes to deal with xarray change
- Now compute mod_i to improve performance
- Added save_interpolation argument
- Some changes to deal with GEOS-Chem files that have multiple species (as for Jacobian construction)
- Create the OBS_DIR/operator_components directory to store the interpolation

Commit:
- Added NUM_THREADS arguments to attempt to improve computational efficiency (this is assumed to be 8 and is also untested)
- Removed reading the config file from this file to allow for different config file names
- I think the rest of the changes are mostly to account for files that have multiple simulated species (i.e., for Jacobian calculation)

Commit:
- Added NUM_THREADS specification to attempt to improve performance. Unclear that this works. Also defaults to 8.
- Changed _open_geoschem and related files function to open files that have multiple species--and specifically, multiple tracers of the same type for Jacobian simulations. This assumes that the tracers have format something like CH4_0xxx (though it's pretty flexible).

Commit:
Removing timing statements and also fixing OCO2 parser

Commit:
- Add NUM_THREADS (may or may not work, defaults to 8)
- Separate out MODEL_LEVEL_EDGE_DIR and MODEL_CONCENTRATION_DIR (again, this is for Jacobian simulations for which these directories are not the same)
- Allow for colocation indices to be loaded or computed and saved. This is often somewhat slow. The indices are saved in the same place as the interpolation files.

Commit:
Early edits to develop TROPOMI parsers

pennelise commented on config.yaml:
```diff
@@ -25,21 +27,16 @@ OCO2_v11.1_preprocessed:
-    PRESSURE_EDGES: 'pressure_edges' # This variable does not
-    # exist in the TROPOMI files and needs to be created in the
-    # parser.
-    # N_CENTERS: 'nlayers'
```
We actually should keep these. The reason is just because we don't want file-specific strings hardcoded inside functions. I'm trying to keep all those strings in this config file so they can be found easily and changed in only one place if the names change. I can add them back.

pennelise commented on config.yaml:
```diff
@@ -49,6 +46,22 @@ TROPOMIvXX: # version is specified for each instrument
+  PARSER: 'read_TROPOMI_blended'
+  DATA_FIELDS: 
+    N_OBS: 'nobs'
+    N_EDGES: 'none'        # These variables do not exist and are  
```
See above. We should add N_CENTERS, DELTA_PRESSURE, etc back.

pennelise commented on interpolation.py:
```diff
@@ -49,14 +59,14 @@ def __expand_profile_dims(self):
         assert (
-            self.model_conc_at_layers.ndim == 2
-        ), "GEOS-Chem methane layers must be 2D (nobs x nlevels), or 1D (nlevels)."
+            self.model_conc_at_layers.ndim >= 2
```
Change to == 3 or ==2 (because I think e.g. 4 dims is gonna create issues)

pennelise replied:
This is a as good as it gets actually. Close it.

pennelise commented on interpolation.py:
```diff
@@ -143,35 +153,53 @@ def get_hprime_satellite_edges(self):
     def interpolate(self):
         """
-        Interpolate GEOS-Chem methane to satellite edges OR centers.
+        Interpolate GEOS-Chem methane to satellite edges OR centers. Use
```
I now agree that we should keep this save-out option.

pennelise commented on interpolation.py:
```diff
@@ -143,35 +153,53 @@ def get_hprime_satellite_edges(self):
 
+        # Get the interpolation map
+        try:
+            interpolation_map = np.load(f"{self.save_dir}_interpolation.npy")
```
is pickle a stable way to save this? I have run into issues with pickle requiring the same version of all your python packages used to save it in order to read it.

pennelise commented on interpolation.py:
```diff
@@ -143,35 +153,53 @@ def get_hprime_satellite_edges(self):
+                )
+            
+            # Save out the interpolation map
+            if self.save_interpolation.lower() == "true":
```
move this inside the except loop to avoid saving out an interpolation map that has already been saved.

pennelise commented on main.py:
```diff
@@ -45,22 +52,32 @@ def apply_operator_to_chunks(model_conc_files,
             operators.get_model_columns(
-                mod_i, sat_i.where(~missing_times, drop=True), 
-                config["LOCAL_SETTINGS"]["SATELLITE_NAME"]))
+                mod_i, sat_i.where(~missing_times, drop=True),
```
I think we may be able to simplify this function call and just input the configs and then parse them inside the function

pennelise commented on parsers.py:
```diff
@@ -52,24 +82,53 @@ def read_satellite_file(file_path, data_fields):
+    # TROPOMI objects.
+    z = satellite.layer.size
+    p_surface = satellite['surface_pressure'].values.reshape((-1, 1))
+    p_interval = satellite['pressure_interval'].values.reshape((-1, 1))
```
these should be defined in the config

hannahnesser replied:
See above discussion.

pennelise commented on utilities.py:
```diff
@@ -104,34 +112,48 @@ def get_missing_times(satellite_times, model_times):
 
 
-def colocate_obs(model, satellite):
+def colocate_obs(model, satellite, save_dir=None):
```
again wondering is maybe we pass around all the local settings...

hannahnesser commented on parsers.py:
```diff
@@ -1,9 +1,41 @@
+                save_vars.update({var_pattern : default_name})
+
+    # Return the subsetted data
+    return data[save_vars.keys()].rename(save_vars)
```
This is needed to handle multiple tracers.

Commit:
More detailed documentation of the config file

Commit:
Added TROPOMI and TCCON operators and the corresponding options to the config file. Also, restored N_CENTERS since it is needed for satellites that have center-based grids

Commit:
Added the option to not expand the model edges. This is really only for the TCCON parser, which uses VerticalGrid to regrid the prior and averaging kernels. Also, a few formatting changes for readability.

Commit:
Simplified some function calls to take the config file instead of individual config variables. Also, fixed dropping variables that have vertical dimensions for the case when the satellite data is saved out along with the model columns

Commit:
Simplified function call to take the config file instead of individual config variables.

Commit:
Added the TROPOMI and TCCON parser and debugged the function that checks for the vertical ordering of the pressure levels

Commit:
Moved the utility function get_satellite_parser into parsers.py, which helps to avoid an import of parsers into utilities. Also, a bug fix for the file format string

Commit:
removed unnecessary comment

hannahnesser committed:
Merge branch 'main' into feature/remove_config

hannahnesser committed:
Remove threading environment variable settings

Remove environment variable settings for threading. Eventually, we need to find a way to dynamically add these in accounting for the number of threads a user has available.

hannahnesser committed:
Remove thread count environment variable settings

Remove environment variable settings for thread counts.

hannahnesser committed:
Remove thread count environment variable settings

Remove environment variable settings for thread counts.

hannahnesser committed:
Remove thread limit environment variable settings

Removed environment variable settings for thread limits.
